src/util/transaction_identifier.h

```diff
@@ -0,0 +1,54 @@
+#ifndef BITCOIN_UTIL_TRANSACTION_IDENTIFIER_H
+#define BITCOIN_UTIL_TRANSACTION_IDENTIFIER_H
+
+#include <uint256.h>
+
+#include <type_traits>
+
+/** transaction_identifier represents the two canonical transaction identifier
+ * types (txid, wtxid). */
+template <bool has_witness>
+class transaction_identifier : public uint256
+{
+    static constexpr bool negated_has_witness =
+        std::negation<std::bool_constant<has_witness>>::value;
+    using opposite_type =
+        transaction_identifier<negated_has_witness>;
+
+public:
+    transaction_identifier() : uint256{} {}
+    explicit transaction_identifier(const uint256& other) : uint256{other} {}
+    explicit transaction_identifier(const uint256&& other) : uint256{other} {}
+
+    // Delete copy/move constructors/assignment for differing transaction_identifier types.
+    transaction_identifier(const opposite_type&) = delete;
+    transaction_identifier(opposite_type&&) = delete;
+    transaction_identifier& operator=(const opposite_type&) = delete;
+    transaction_identifier& operator=(opposite_type&&) = delete;
+
+    // Disallow comparisons between different transaction_identifier types.
+    friend bool operator==(const transaction_identifier<has_witness>& a,
+                           const opposite_type& b) = delete;
+    friend bool operator!=(const transaction_identifier<has_witness>& a,
+                           const opposite_type& b) = delete;
+    friend bool operator<(const transaction_identifier<has_witness>& a,
+                          const opposite_type& b) = delete;
+    int Compare(const opposite_type&) const = delete;
+
+    // Allow comparisons between the same transaction_identifier type.
+    friend bool operator==(const transaction_identifier<has_witness>& a,
+                           const transaction_identifier<has_witness>& b) { return a.Compare(b) == 0; }
+    friend bool operator!=(const transaction_identifier<has_witness>& a,
+                           const transaction_identifier<has_witness>& b) { return a.Compare(b) != 0; }
+    friend bool operator<(const transaction_identifier<has_witness>& a,
+                          const transaction_identifier<has_witness>& b) { return a.Compare(b) < 0; }
+    constexpr int Compare(const transaction_identifier<has_witness>& other) const { return static_cast<const uint256&>(*this).Compare(other); }
+};
+
+/** Txid commits to all transaction fields except the witness. */
+using Txid = transaction_identifier<false>;
+/** Wtxid commits to all transaction fields including the witness. */
+using Wtxid = transaction_identifier<true>;
```

Comment:
Is it not possible to replace all of that with just:

```cpp
class Txid : public uint256
{
};

class Wtxid : public uint256
{
};
```

It achieves the purpose of not allowing conversion and comparison between `Txid` and `Wtxid` and each one of them can be treated as `uint256`.

Reply:
That is what I originally had in this PR but it actually doesn't achieve the same on all fronts, so I decided to make it explicit to avoid subtle bugs.

e.g. the following was still permitted:
```c++
Txid txid;
Wtxid wtxid{txid};
// or
wtxid.Compare(txid);
// or
bool equal = (wtxid == txid); // same for !=
```